### ISSUE TITLE
iOS: Restoring non-consumable purchase for sandbox user doesn't work consistently

### ISSUE THREAD

**Issue by WieFel** [RevenueCat/purchases-flutter#224]
**Describe the bug**
I am currently testing my IAP for a non-consumable product on iOS. On Android, it already works fine.

When my app starts, I...
1. Setup RevenueCat
2. Get the purchaser info and
3. Restore transactions for previous IAPs

When running with my sandbox user on my iPhone SE, I experience a little strange behaviour. Assume that the non-consumable was previously purchased using the sandbox user. I use anonymous user IDs.
Then:
1. When I freshly install the app on the phone (completely removed it from the phone before), the restoring of the purchased non-consumable works just fine. It is restored correctly.
Debug logs:
```
[Purchases] - DEBUG: ℹ️ Debug logging enabled
[Purchases] - DEBUG: ℹ️ SDK Version - 3.12.2
[Purchases] - DEBUG: 👤 Initial App User ID - (null)
[Purchases] - DEBUG: 👤 Identifying App User ID: $RCAnonymousID:19734ce4f3704614831616b65d066f22
[Purchases] - DEBUG: ℹ️ Delegate set
[Purchases] - DEBUG: ℹ️ There are no requests currently running, starting request GET /subscribers/$RCAnonymousID0X0P+019734ce4f3704614831616b65d066f22
[Purchases] - DEBUG: ℹ️ API request started: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22
[Purchases] - DEBUG: ℹ️ API request started: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22/offerings
[Purchases] - DEBUG: ℹ️ PurchaserInfo cache is stale, updating from network in foreground.
[Purchases] - DEBUG: 😻 PurchaserInfo updated from network.
[Purchases] - DEBUG: ℹ️ API request completed with status: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22/offerings 200
[Purchases] - DEBUG: ℹ️ Requesting products from the store with identifiers: {(
    "ads_remove"
)}
[Purchases] - DEBUG: ℹ️ Products request finished.
[Purchases] - DEBUG: 💰 Retrieved SKProducts: 
[Purchases] - DEBUG: 💰 ads_remove - <SKProduct: 0x281b624d0>
[Purchases] - DEBUG: ℹ️ 1 completion handlers waiting on products
[Purchases] - DEBUG: ℹ️ API request completed with status: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22 201
[Purchases] - DEBUG: ℹ️ Sending latest PurchaserInfo to delegate.
[Purchases] - DEBUG: ℹ️ Serial request done: GET /subscribers/$RCAnonymousID0X0P+019734ce4f3704614831616b65d066f22, 0 requests left in the queue
[Purchases] - DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
[Purchases] - DEBUG: ℹ️ Loaded receipt from url file:///private/var/mobile/Containers/Data/Application/24658B2C-6A95-4B31-A3FA-662C5915D7BD/StoreKit/sandboxReceipt
[Purchases] - INFO: ℹ️ Parsing receipt
[Purchases] - WARN: ⚠️ /Users/felix/Desktop/bleckl/ios/Pods/PurchasesCoreSwift/PurchasesCoreSwift/LocalReceiptParsing/ReceiptParser.swift-receiptHasTransactions(receiptData:): Could not parse receipt, conservatively returning true
[Purchases] - DEBUG: ℹ️ Found 0 unsynced attributes for App User ID: $RCAnonymousID:19734ce4f3704614831616b65d066f22
[Purchases] - DEBUG: ℹ️ There are no requests currently running, starting request POST /receipts
[Purchases] - DEBUG: ℹ️ API request started: POST /v1/receipts
[Purchases] - DEBUG: ℹ️ API request completed with status: POST /v1/receipts 200
[Purchases] - DEBUG: ℹ️ Serial request done: POST /receipts, 0 requests left in the queue
[Purchases] - DEBUG: ℹ️ Sending updated PurchaserInfo to delegate.
```

2. However, when I hit hot-reload (or stop the app and re-run it, without uninstalling it first), suddenly the previous purchase isn't restored at all.
Debug logs:
```
[Purchases] - DEBUG: ℹ️ Debug logging enabled
[Purchases] - DEBUG: ℹ️ SDK Version - 3.12.2
[Purchases] - DEBUG: 👤 Initial App User ID - (null)
[Purchases] - INFO: ℹ️ Purchases instance already set. Did you mean to configure two Purchases objects?
[Purchases] - DEBUG: ℹ️ Sending latest PurchaserInfo to delegate.
[Purchases] - DEBUG: ℹ️ Delegate set
[Purchases] - DEBUG: ℹ️ There are no requests currently running, starting request GET /subscribers/$RCAnonymousID0X0P+019734ce4f3704614831616b65d066f22
[Purchases] - DEBUG: ℹ️ API request started: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22
[Purchases] - DEBUG: ℹ️ API request started: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22/offerings
[Purchases] - DEBUG: ℹ️ Vending PurchaserInfo from cache.
[Purchases] - DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
[Purchases] - DEBUG: ℹ️ API request completed with status: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22/offerings 304
[Purchases] - DEBUG: ℹ️ Requesting products from the store with identifiers: {(
    "ads_remove"
)}
[Purchases] - DEBUG: ℹ️ API request completed with status: GET /v1/subscribers/$RCAnonymousID:19734ce4f3704614831616b65d066f22 200
[Purchases] - DEBUG: ℹ️ Serial request done: GET /subscribers/$RCAnonymousID0X0P+019734ce4f3704614831616b65d066f22, 0 requests left in the queue
[Purchases] - DEBUG: ℹ️ Products request finished.
[Purchases] - DEBUG: 💰 Retrieved SKProducts: 
[Purchases] - DEBUG: 💰 ads_remove - <SKProduct: 0x281b4b770>
[Purchases] - DEBUG: ℹ️ 1 completion handlers waiting on products
[Purchases] - DEBUG: ℹ️ Loaded receipt from url file:///private/var/mobile/Containers/Data/Application/24658B2C-6A95-4B31-A3FA-662C5915D7BD/StoreKit/sandboxReceipt
[Purchases] - INFO: ℹ️ Parsing receipt
[Purchases] - WARN: ⚠️ /Users/felix/Desktop/bleckl/ios/Pods/PurchasesCoreSwift/PurchasesCoreSwift/LocalReceiptParsing/ReceiptParser.swift-receiptHasTransactions(receiptData:): Could not parse receipt, conservatively returning true
[Purchases] - DEBUG: ℹ️ Found 0 unsynced attributes for App User ID: $RCAnonymousID:19734ce4f3704614831616b65d066f22
[Purchases] - DEBUG: ℹ️ There are no requests currently running, starting request POST /receipts
[Purchases] - DEBUG: ℹ️ API request started: POST /v1/receipts
[Purchases] - DEBUG: ℹ️ API request completed with status: POST /v1/receipts 200
[Purchases] - DEBUG: ℹ️ Serial request done: POST /receipts, 0 requests left in the queue
[Purchases] - DEBUG: ℹ️ Vending Offerings from cache
```

I don't know what could be causing this issue... Apparently, the user ID is the same for both, the fresh install and the re-run of the app. From that point of view, everything is correct...
Any ideas?

I am using `purchases_flutter: ^3.4.0` and flutter version `2.2.3`.

**Comment by aboedo** [comment]
Hi! Thanks for opening! 

Could you elaborate on the actual use case? 

Are you setting up a restore button? Are you manually calling the restore methods at all? Or just calling setup? Does the entitlement correctly show up for the user in the dashboard? Is it accessible through `purchaserInfo`?

**Comment by aboedo** [comment]
I hadn't noticed this at first, but from the logs from the hot-reload: 

> [Purchases] - INFO: ℹ️ Purchases instance already set. Did you mean to configure two Purchases objects?

This message is issued whenever `configure` is called but the SDK had already been configured. This can lead to issues, and might be the cause of the unexpected behavior you're experiencing here.

**Comment by WieFel** [comment]
@aboedo thanks for your answers. Maybe I am doing something wrong in the initialization. Maybe you can help me if I tell you how my app is set up:
On app startup (in the splash screen) I initialize RevenueCat by calling the following function:
```dart
  Future<void> _initRevenueCat() async {
    await Purchases.setDebugLogsEnabled(true);
    await Purchases.setup("<MY_API_KEY>");
    await _purchaseProvider.init();
  }
```
where `_purchaseProvider.init()` does the following:
1. It registers a purchaser info listener by calling `Purchases.addPurchaserInfoUpdateListener(_onPurchaseInfoUpdate)`, where `_onPurchaseInfoUpdate` looks as follows:
```dart
  void _onPurchaseInfoUpdate(PurchaserInfo info) {
    if (info.entitlements.all[REMOVE_ADS_PRODUCT_ID]?.isActive ?? false) {
      // Grant user "pro" access
      hasPurchasedRemoveAds = true;
    }
  }
```
2. It retrieves purchaser info and tries to restore transactions by calling `_getSubscriptionInfo()` which looks as follows:
```dart
  Future<void> _getSubscriptionInfo() async {
    try {
      await Purchases.getPurchaserInfo();
      await Purchases.restoreTransactions();
    } on PlatformException catch (_) {
      // Error fetching purchaser info -> do nothing, flag is already false
    }
  }
```
I listen for the results of `getPurchaserInfo` and `restoreTransactions` inside the registered listener of (1). Not sure if it's a bad idea to do it that way!?

**Comment by aboedo** [comment]
@WieFel sorry for the delayed response. 

I think your code is fine in general, just make sure that wherever you're calling `setup` is only called once. It seems like it's getting called again on hot reloads, and this creates a new instance of purchases, but the old once is still loaded into memory. 

For your code, though, I'd just clarify that the SDK already loads `purchaserInfo` automatically as soon as it starts up, so the call to `await Purchases.getPurchaserInfo();` is unnecessary.

**Comment by WieFel** [comment]
> I think your code is fine in general, just make sure that wherever you're calling `setup` is only called once.

Thank you very much, @aboedo! Is there any way I can check whether purchases is already set up, so that I can avoid re-calling `setup`?

> For your code, though, I'd just clarify that the SDK already loads `purchaserInfo` automatically as soon as it starts up, so the call to `await Purchases.getPurchaserInfo();` is unnecessary. 

Oh, also good to know, thanks! So basically the only thing I need to do manually is restoring transactions, right!?

**Comment by aboedo** [comment]
> Oh, also good to know, thanks! So basically the only thing I need to do manually is restoring transactions, right!?

yes! and that is only necessary if you're not using an accounts system. 

> Thank you very much, @aboedo! Is there any way I can check whether purchases is already set up, so that I can avoid re-calling setup?

There's a new property in `purchases-ios`, but it hasn't made its way into `purchases-flutter` yet 😞 

as a proxy for it, you could try using the `appUserID` property. If it's null or an error is thrown when trying to access it, the SDK hasn't been configured. 

so something like
```dart
var isPurchasesConfigured = false;
try {
    isPurchasesConfigured = Purchases.appUserID != null;
} on PlatformException catch (_) {
    // Purchases instance hasn't been configured yet
}
```

**Comment by aboedo** [comment]
version 3.4.1 is out with the fix!! Closing this new, feel free to open a new issue if you run into anything else!!

**Comment by aboedo** [comment]
did you ever figure this out?

**Comment by WieFel** [comment]
Will report back when I get to it, the next days ;)

**Comment by WieFel** [comment]
@aboedo thanks for your help! Now I could solve the problem as follows:

1. I add the listener for purchaser info updates already **before** calling `Purchases.setup`:
```dart
...
Purchases.addPurchaserInfoUpdateListener(<my_listener>);
await Purchases.setup("<MY_API_KEY>");
...
```
2. I removed the manual call of `await Purchases.getPurchaserInfo();` and only kept the call for `await Purchases.restoreTransactions();` in place.

I didn't even need to do this
> so something like
> 
> ```dart
> var isPurchasesConfigured = false;
> try {
>     isPurchasesConfigured = Purchases.appUserID != null;
> } on PlatformException catch (_) {
>     // Purchases instance hasn't been configured yet
> }
> ```

Issue can be closed :)

**Comment by aboedo** [comment]
@WieFel awesome!! 🎉 
thanks for reporting back! let us know if you run into any more issues. Closing this now. Have a great weekend!
